share/completions/istioctl.fish

```diff
@@ -0,0 +1,176 @@
+    __istioctl_debug "DirectiveLine: $directiveLine"
+    __istioctl_debug "flagPrefix: $flagPrefix"
+
+    for comp in $comps
+        printf "%s%s\n" "$flagPrefix" "$comp"
+    end
+
+    printf "%s\n" "$directiveLine"
+end
+
+# This function does two things:
+# - Obtain the completions and store them in the global __istioctl_comp_results
+# - Return false if file completion should be performed
+function __istioctl_prepare_completions
+    __istioctl_debug ""
+    __istioctl_debug "========= starting completion logic =========="
+
+    # Start fresh
+    set --erase __istioctl_comp_results
+
+    set -l results (__istioctl_perform_completion)
+    __istioctl_debug "Completion results: $results"
+
+    if test -z "$results"
+        __istioctl_debug "No completion, probably due to a failure"
+        # Might as well do file completion, in case it helps
+        return 1
+    end
+
+    set -l directive (string sub --start 2 $results[-1])
+    set --global __istioctl_comp_results $results[1..-2]
+
+    __istioctl_debug "Completions are: $__istioctl_comp_results"
+    __istioctl_debug "Directive is: $directive"
+
+    set -l shellCompDirectiveError 1
+    set -l shellCompDirectiveNoSpace 2
+    set -l shellCompDirectiveNoFileComp 4
+    set -l shellCompDirectiveFilterFileExt 8
+    set -l shellCompDirectiveFilterDirs 16
+
+    if test -z "$directive"
+        set directive 0
+    end
+
+    set -l compErr (math (math --scale 0 $directive / $shellCompDirectiveError) % 2)
+    if test $compErr -eq 1
+        __istioctl_debug "Received error directive: aborting."
+        # Might as well do file completion, in case it helps
+        return 1
+    end
+
+    set -l filefilter (math (math --scale 0 $directive / $shellCompDirectiveFilterFileExt) % 2)
+    set -l dirfilter (math (math --scale 0 $directive / $shellCompDirectiveFilterDirs) % 2)
+    if test $filefilter -eq 1; or test $dirfilter -eq 1
+        __istioctl_debug "File extension filtering or directory filtering not supported"
+        # Do full file completion instead
+        return 1
+    end
+
+    set -l nospace (math (math --scale 0 $directive / $shellCompDirectiveNoSpace) % 2)
+    set -l nofiles (math (math --scale 0 $directive / $shellCompDirectiveNoFileComp) % 2)
+
+    __istioctl_debug "nospace: $nospace, nofiles: $nofiles"
+
+    # If we want to prevent a space, or if file completion is NOT disabled,
+    # we need to count the number of valid completions.
+    # To do so, we will filter on prefix as the completions we have received
+    # may not already be filtered so as to allow fish to match on different
+    # criteria than the prefix.
+    if test $nospace -ne 0; or test $nofiles -eq 0
+        set -l prefix (commandline -t | string escape --style=regex)
+        __istioctl_debug "prefix: $prefix"
+
+        set -l completions (string match -r -- "^$prefix.*" $__istioctl_comp_results)
+        set --global __istioctl_comp_results $completions
```

Comment:
This global variable is quite suspect. This function should simply print the completion candidates instead of setting a global variable.